src/libraries/System.Private.CoreLib/src/System/Reflection/Nullability/NullabilityInfoContext.cs

```diff
@@ -0,0 +1,430 @@
+                {
+                    return new NullabilityInfo(fieldInfo.FieldType, NullabilityState.Unknown, NullabilityState.Unknown, null, s_emptyArray);
+                }
+            }
+
+            var attributes = fieldInfo.GetCustomAttributesData();
+            var nullability = GetNullabilityInfo(fieldInfo, fieldInfo.FieldType, attributes);
+            CheckNullabilityAttributes(nullability, attributes);
+            return nullability;
+        }
+
+        private bool IsPublicOnly(bool isPrivate, bool isFamilyAndAssembly, bool isAssembly, Module module)
+        {
+            if (!_publicOnlyModules.TryGetValue(module, out var value))
+            {
+                value = PopulateAnnotationInfo(module.GetCustomAttributesData());
+                _publicOnlyModules.Add(module, value);
+            }
+
+            if (value == NotAnnotatedStatus.None)
+            {
+                return false;
+            }
+
+            if ((isPrivate || isFamilyAndAssembly) && value.HasFlag(NotAnnotatedStatus.Private) ||
+                 isAssembly && value.HasFlag(NotAnnotatedStatus.Internal))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private NotAnnotatedStatus PopulateAnnotationInfo(IList<CustomAttributeData> customAttributes)
+        {
+            foreach (var attribute in customAttributes)
+            {
+                if (attribute.AttributeType.Name == "NullablePublicOnlyAttribute" &&
+                    attribute.AttributeType.Namespace == CompilerServicesNameSpace &&
+                    attribute.ConstructorArguments.Count == 1)
+                {
+                    if (attribute.ConstructorArguments[0].Value is bool boolValue && boolValue)
+                    {
+                        return NotAnnotatedStatus.Internal | NotAnnotatedStatus.Private;
+                    }
+                    else
+                    {
+                        return NotAnnotatedStatus.Private;
+                    }
+                }
+            }
+
+            return NotAnnotatedStatus.None;
+        }
+
+        [RequiresUnreferencedCode("Nullability attributes are trimmed by the linker")]
+        private NullabilityInfo GetNullabilityInfo(MemberInfo memberInfo, Type type, IList<CustomAttributeData> customAttributes) =>
+            GetNullabilityInfo(memberInfo, type, customAttributes, 0);
+
+        [RequiresUnreferencedCode("Nullability attributes are trimmed by the linker")]
+        private NullabilityInfo GetNullabilityInfo(MemberInfo memberInfo, Type type, IList<CustomAttributeData> customAttributes, int index)
+        {
+            NullabilityState state = NullabilityState.Unknown;
+
+            if (type.IsValueType)
+            {
+                var underlyingType = Nullable.GetUnderlyingType(type);
+                if (underlyingType != null)
+                {
+                    type = underlyingType;
```

Comment:
Not sure if we should set here the underlying type or the Nullable<T> type for nullable value type.

CC @terrajobst

Reply:
IMO the actual `Nullable<T>` value would be more "correct"; the user can always ask for the underlying type.